src/kudu/consensus/consensus_queue.cc

```diff
@@ -209,6 +209,44 @@ PeerMessageQueue::PeerMessageQueue(const scoped_refptr<MetricEntity>& metric_ent
   log_cache_.Init(queue_state_.last_appended);
 }
 
+void PeerMessageQueue::SetProxyFailureThreshold(
+    int32_t proxy_failure_threshold_ms) {
+  std::lock_guard<simple_spinlock> lock(queue_lock_);
+  proxy_failure_threshold_ms_ = proxy_failure_threshold_ms;
+}
+
+void PeerMessageQueue::SetProxyFailureThresholdLag(
+    int32_t proxy_failure_threshold_lag) {
+  std::lock_guard<simple_spinlock> lock(queue_lock_);
+  proxy_failure_threshold_lag_ = proxy_failure_threshold_lag;
+}
+
+bool PeerMessageQueue::HasProxyPeerFailedUnlocked(
+    const TrackedPeer* proxy_peer, const TrackedPeer* dest_peer) {
+  auto max_proxy_failure_threshold =
+    MonoDelta::FromMilliseconds(proxy_failure_threshold_ms_);
+
+  if (MonoTime::Now() - proxy_peer->last_communication_time >
+      max_proxy_failure_threshold) {
+    // The leader has not communicated with proxy_peer within the
+    // proxy_failure_threshold_ms. Hence this peer cannot act as a 'proxy peer'
+    // and is considered failed
+    return true;
+  }
+
+  bool is_proxy_lagging = (dest_peer->next_index > proxy_peer->next_index) &&
```

Comment:
is there any ramifications of this if dest is behind the proxy and we turn on proxy
when the proxy instance is also lagging by a whole lot?

Reply:
Destination being behind the proxy seems fine, since that is what we want (proxy node should have the ops needed to build the request for the destination). Other case is the problematic one: proxy node is behind the destination node. This check will route directly to destination until proxy has caught up sufficiently.